### ISSUE TITLE
Stale listener condition present in Gateway for a non-existing Listener

### ISSUE THREAD

**Issue by chauhanshubham** [envoyproxy/gateway#773]
*Description*:
Following the reproduction steps, it is seen that the Gateway is left with a stale listener condition, for a listener
that no longer exists.

*Repro steps*:
1. apply `examples/kubernetes/quickstart.yaml` - ensure that Gateway has proper Status conditions, with 1 listener condition
showing `attachedRoutes: 1`.
2. Add another listener in the gateway, for the listener spec to look something like this
```yaml
listeners:
- allowedRoutes:
     namespaces:
       from: Same
     name: http
     port: 8080
     protocol: HTTP
- allowedRoutes:
     namespaces:
       from: Same
     name: http2
     port: 8081
     protocol: HTTP 
```
This should create 2 listener conditions, both specifying `attachedRoutes: 1`.
3. Remove the single HTTPRoute that was created by the `quickstart.yaml`. Both existing listener conditions are updated correctly, specifying `attachedRoutes: 0`.
4. Re-apply the `quickstart.yaml` which will "recreate" the HTTPRoute, and "update" the Gateway to have only one listener.
5. The bug is that - in the Gateway status still shows the stale listener Condition corresponding to listener `http2` with `attachedRoutes: 0`
The spec and status thus looks like this
```yaml
 spec:
    gatewayClassName: eg
    listeners:
    - allowedRoutes:
        namespaces:
          from: Same
      name: http
      port: 8080
      protocol: HTTP
  status:
    ....
    conditions:
    ....
    listeners:
    - attachedRoutes: 1
      conditions:
      - lastTransitionTime: "2022-12-03T06:58:58Z"
        message: Listener is ready
        observedGeneration: 3
        reason: Ready
        status: "True"
        type: Ready
      name: http
      supportedKinds:
      - group: gateway.networking.k8s.io
        kind: HTTPRoute
    - attachedRoutes: 0
      conditions:
      - lastTransitionTime: "2022-12-03T06:58:48Z"
        message: Listener is ready
        observedGeneration: 2
        reason: Ready
        status: "True"
        type: Ready
      name: http2
      supportedKinds:
      - group: gateway.networking.k8s.io
        kind: HTTPRoute
```

**Comment by arkodg** [comment]
thanks for raising this issue, there are two ways to fix this issue
* either we always reset the listener status before computing it in https://github.com/envoyproxy/gateway/blob/7b73d14177df608444e085f46abd2073e44671a3/internal/gatewayapi/translator.go#L189
* or once we are done computing newer listener status, we parse through all listener status and remove the stale ones

**Comment by haq204** [comment]
Hi @arkodg I can take on this issue since I have another issue currently blocked by the 0.6.0 gateway-api bump.

From a brief scan, it seems like the easiest way is to reset the listener status before computing. It doesn't look like `ProcessListeners` is responsible for communicating to the API server for updating the status which was my initial concern but I think it seems safe to do because the gatewayapi translator will always update the status for all gateways for a given ingress class when it receives an event from the kubernetes provider for gateway api resource changes though I could be wrong on that last part.

**Comment by arkodg** [comment]
yeah @haq204 the easiest/cleanest approach is the one you suggested, to just reset the status conditions before starting translation in the Gateway API library. It would look similar to https://github.com/envoyproxy/gateway/blob/97032bee2db5ec73027b172e24b03d4dfe98ac3a/internal/gatewayapi/translator.go#L134

**Comment by arkodg** [comment]
we also need to ensure that resetting and recomputing the status doesnt end up updating the status condition in K8s if the condition is same (all fields except `LastTransitionTime` is same)
this is taken care of today here https://github.com/envoyproxy/gateway/blob/97032bee2db5ec73027b172e24b03d4dfe98ac3a/internal/status/status.go#L163 but we need to verify that is working as intended

**Comment by danehans** [comment]
@haq204 thanks for taking on this issue. Is the issue still TODO or have you made any progress with it?

**Comment by haq204** [comment]
@danehans still TODO but will try to tackle it this week
